rfc/rfc-77/rfc-77.md

```diff
@@ -0,0 +1,247 @@
+
+Let's take an example where a row's `secondary-key` column is updated from its old value of 
+`old-secondary-key -> record-key` to `new-secondary-key -> record-key`. Note that as the secondary keys are updated, 
+the `new-secondary-key` could get mapped to a different FileGroup (in MDT partition hosting the SI) as compared to 
+`old-secondary-key`. There needs to be a mechanism to clean up the `old-secondary-key -> record_key` mapping (from the 
+MDT partition hosting SI). The proposal is to emit a tombstone entry `old-secondary-key -> (record-key ,deleted)`. This 
+is then followed by writing the `new-secondary-key -> record-key`. The log record reader (merger) and compactor should 
+correctly use the tombstone marker to remove stale entries from the old FileGroup. Note that since the mapping of a 
+record (of the MDT) to a FileGroup is based on the `secondary-key`, it is guaranteed that 
+`old-secondary-key -> (record-key, deleted)` tombstone record will get mapped to the FileGroup containing the 
+`old-secondary-key, record-key` entry.
+
+Another key observation here is that `old-secondary-key` is required to construct the tombstone record. Unlike some other 
+data systems, Hudi does not read the old-image of a row on updates until a merge is executed. It detects that a row is getting updated by simply 
+reading the index and appending the updates in log files. Hence, there needs to be a mechanism to extract `old-secondary-key`. We propose
+`old-secondary-key` to be extracted by scanning the MDT partition (hosting the SI) and doing a reverse lookup based 
+on the `record-key` of the row being updated. It should be noted that this is going to be expensive operation as the 
+base table grows in size (which inherently means that SI will grow in size) in terms of number of rows. One way to 
+optimize this is to build a reverse mapping `record-key -> secondary-key` in a different MDT partition. This is 
+left as a TBD (as of this writing).
+
+##### Deletes (on the base table)
+This is handled the same way updates are handled. Only `old-secondary-key -> (record-key, deleted)` tuple is emitted 
+as a tombstone record to the FileGroup containing the `old-secondary-key -> record-key` entry.
+
+### Query Execution
+Secondary index will be used to prune the candidate files (to scan) if the query has  an "EQUAL TO" or "IN" predicate
+on `secondary-key` columns. The initial implementation will follow the same logic (and constraints) used by RLI based 
+data skipping(like single column keys, rule-based index selection etc.). In the first step, we scan the SI partition
+(in MDT) and filter all entries that match the secondary keys specified in the query predicate. We collect all the 
+record keys (after filtering). Subsequently, these record keys are used to look up the record location using RLI.
+
+### Merging of secondary index records (on the read-path or during compaction)
```

Comment:
@vinothchandar can you please review this part?

Reply:
Updated more details with API changes.

Reply:
I would not overload or extend `HoodieKey` since thats used in the write path. Lets make a new `MergeKey` class for this, which can be wrapping a `HoodieKey` or a `string` for the case where key is just the recordkey or sth.